File: faang/SlidingWindow.java
import java.util.*;

public class Solution {
    public int solve(String string) {

        int len_string = string.length();
        if(len_string == 0 ||  len_string == 1){
            return len_string;
        }
        
        int max_len = 0;

        int p1 = 0;
        int p2 = 0;

        HashMap<String, Integer> map = new HashMap<String, Integer>();


        while(p2 < len_string){
            String letter = string.substring(p2, p2 + 1);
            if(map.containsKey(letter)){
                // already seen
                int actual_len = p2 - p1;
                // sliding pointer
                if (p1 > map.get(letter)){
                    // this is previous data info
                    p1 = p2;
                    // sum 1 for offset
                    actual_len += 1;
                }
                else{
                    p1 = map.get(letter) + 1;
                }

                if(actual_len > max_len)
                    max_len = actual_len;
            }
            map.put(letter, p2);
            // add or update found positions
            p2 += 1;
            
        }

            

        return max_len;
        
    }

    public void test(LinkedList<String> testCases, LinkedList<Integer> expected){

        for(int i=0; i<testCases.size(); i++){
            int answer = new Solution().solve(testCases.get(i));
            if(answer == expected.get(i)){
                System.out.println("OK");
            }
            else{
                System.out.println("Error: for test case " + (i+1) + ", expected: " + expected.get(i) + ", solved: " + answer);
            }
        }
        

    }
    
    public static void main(String[] args){
        LinkedList<String> testCases = new LinkedList<String>();
        LinkedList<Integer> expected = new LinkedList<Integer>();
        
        testCases.add("abcbda");
        testCases.add("abccabb");
        testCases.add("cccccc");
        testCases.add("");
        
        expected.add(4);
        expected.add(3);
        expected.add(1);
        expected.add(0);
        
        
        new Solution().test(testCases, expected);
    }
}

File: faang/SlidingWindow.py


def solve(string):

    len_string = len(string)
    if len_string in [0, 1]:
        return len_string
    
    max_len = 0

    p1 = 0
    p2 = 0
    map = dict()

    while p2 < len_string:
        letter = string[p2]


        if letter in map:
            # already seen
            actual_len = p2 - p1
            # sliding pointer
            if p1 > map[letter]:
                # this is previous data info
                p1 = p2
                # sum 1 for offset
                actual_len += 1
            else:
                # we see in this sequence
                p1 = map[letter] + 1

            if actual_len > max_len:
                max_len = actual_len

        map[letter] = p2
        # add or update found positions
        p2 += 1

    return max_len

def test():

    test_cases = [
        "abcbda",
        "abccabb",
        "cccccc",
        "",
    ]
    expected = [
        4,
        3,
        1,
        0,
    ]
    for i in range(len(test_cases)):
        solution = None
        try:
            solution = solve(test_cases[i])
            assert solution == expected[i]
            print("OK")
        except Exception as error:
            print(f"Error, test {test_cases[i]}, expected {expected[i]}, calculated {solution}", error)

test()